### PR TITLE
pprof: fix incorrect parsing for locations

### LINKED ISSUE #61: could not parse raw pprof output: malformed location line: 1: 0x60671d M=1
I was trying out go-torch with one of our [CockroachDB](https://github.com/cockroachdb/cockroach) test clusters. I'm running OSX and the server is running Ubuntu 16.04.

go-torch at 55a555ddd4d6656c3b55509025681ceb9ff1e154.

With the following file (obtained via `go tool pprof -raw -seconds 5 http://localhost:8080/debug/pprof/profile > profile.raw`)

[torch.pb.gz](https://github.com/uber/go-torch/files/1132993/torch.pb.gz)

I get

```
$ go-torch -binaryinput torch.pb.gz
INFO[09:46:19] Run pprof command: go tool pprof -raw -seconds 30 torch.pb.gz
FATAL[09:46:19] Failed: could not parse raw pprof output: malformed location line: 183: 0x16e11e0 M=1
```

Note the above is the distilled repro - I initially tried

`go-torch -t 5 -u https+insecure://<host>:8080/debug/pprof/profile` and also 
[pprof.txt](https://github.com/uber/go-torch/files/1132998/pprof.txt) + `go-torch -t 5` but the result is the same - I always get an error of the form

> Failed: could not parse raw pprof output: malformed location line: XXX: 0xDEADBEEF M=\d

Running pprof directly everything looks good-ish:

```
$ go tool pprof -seconds 5 http://localhost:8080/debug/pprof/profile
Fetching profile from http://localhost:8080/debug/pprof/profile?seconds=5
Please wait... (5s)
Saved profile in /Users/tschottdorf/pprof/pprof.cockroach.linux-gnu-amd64.6ba72ad700340e95483d4d96000a141e3d528f84.localhost:8080.samples.cpu.015.pb.gz
Entering interactive mode (type "help" for commands)
(pprof) top5
2980ms of 9290ms total (32.08%)
Dropped 709 nodes (cum <= 46.45ms)
Showing top 5 nodes out of 380 (cum >= 410ms)
      flat  flat%   sum%        cum   cum%
     760ms  8.18%  8.18%     1000ms 10.76%  [cockroach.linux-gnu-amd64.6ba72ad700340e95483d4d96000a141e3d528f84]
     750ms  8.07% 16.25%     1730ms 18.62%  runtime.scanobject
     620ms  6.67% 22.93%      620ms  6.67%  runtime.heapBitsForObject
     440ms  4.74% 27.66%      460ms  4.95%  syscall.Syscall
     410ms  4.41% 32.08%      410ms  4.41%  runtime.greyobject
```

(and you get a real call graph in `web`, i.e. there isn't anything interesting in the above list because the binary is mostly idle: [torch.pdf](https://github.com/uber/go-torch/files/1133009/torch.pdf)
), but it doesn't know how to associate the source files:

```
(pprof) list hlc
Total: 9.63s
No source information for github.com/cockroachdb/cockroach/pkg/util/hlc.(*Clock).Now
No source information for github.com/cockroachdb/cockroach/pkg/util/hlc.(*Timestamp).Size
No source information for github.com/cockroachdb/cockroach/pkg/util/hlc.(*Timestamp).String
No source information for github.com/cockroachdb/cockroach/pkg/util/hlc.Timestamp.String
No source information for github.com/cockroachdb/cockroach/pkg/util/hlc.sovTimestamp
```

Re-running this (with the `cockroach` binary downloaded from the remote machine) gives better results:

```
$ go tool pprof -seconds 5 cockroach http://localhost:8080/debug/pprof/profile
pprof) list hlc
Total: 9.93s
ROUTINE ======================== github.com/cockroachdb/cockroach/pkg/util/hlc.(*Timestamp).MarshalTo in /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/util/hlc/timestamp.pb.go
      10ms       10ms (flat, cum)   0.1% of Total
         .          .     68:	i++
         .          .     69:	i = encodeVarintTimestamp(dAtA, i, uint64(m.WallTime))
         .          .     70:	dAtA[i] = 0x10
         .          .     71:	i++
         .          .     72:	i = encodeVarintTimestamp(dAtA, i, uint64(m.Logical))
      10ms       10ms     73:	return i, nil
         .          .     74:}
         .          .     75:
         .          .     76:func encodeFixed64Timestamp(dAtA []byte, offset int, v uint64) int {
         .          .     77:	dAtA[offset] = uint8(v)
         .          .     78:	dAtA[offset+1] = uint8(v >> 8)
[...]
```

I'm not sure how relevant that is, but I've often found it confusing to use pprof across systems, so perhaps there's a clue there. Also happy to privately share the server name if someone wants to test against the real thing (though the saved profile is probably all you need).

### PR DESCRIPTION
**Pull request by irfansharif**
Fixes #61.

Extrapolating from
https://github.com/google/pprof/blob/master/profile/profile.go#L360 and
https://github.com/uber/go-torch/issues/61 we can have locations of the
following format:

    <id>: 0x<address> M=<mapping id>

Note: This doesn't address all malformed inputs, malformed inputs that
may have gotten through before will still do so. This patch only
addresses #61.

Added the raw pprof out as an additional test for one of our binaries
that failed prior.

### CODE REVIEW

**Comment by tschottdorf** [comment]
Nice, thanks @irfansharif!

File: pprof/parser.go [hunk 1]
```diff
@@ -162,10 +162,12 @@ func (p *rawParser) toProfile() (*stack.Profile, error) {
 func (p *rawParser) addLocation(line string) {
 	parts := splitBySpace(line)
 	if len(parts) < 4 {
-		// Do not error when there is only id and address in the line.
-		if len(parts) != 2 {
-			p.setError(fmt.Errorf("malformed location line: %v", line))
+		if len(parts) == 2 || len(parts) == 3 && strings.HasPrefix(parts[2], "M=") {
```
Comment (GautamGupta) [inline pprof/parser.go]: Would recommend parenthesis

Comment (prashantv) [inline pprof/parser.go]: Can we clean this up into separate cases:
```go
if len(parts) < 4 {
  switch {
  case len(parts) == 2:
    // some lines just have an ID and address, we can ignore those
    return
  case len(parts) == 3 && strings.HasPrefix(...):
    // [...]
    return
  }
  // malformed location line error
}
```

Comment (irfansharif) [inline pprof/parser.go]: sure, done.

**Comment by irfansharif** [comment]
> Can you please add a unit test to parser_test.go as well? 

 done.

**Comment by prashantv** [comment]
Just looked at `TestParseSampleCountMismatch` -- it's not checking the actual number of samples, but the count of values vs headers for the sample. E.g., we have `samples/count` `cpu/nanoseconds` and `allocobjects/count`, but we're only specifying 2 sample values, `2` and `10000000`. If you change `2   10000000: 1` to `2   10000000  3: 1`, the test fails as expected.

So that test seems fine, thanks for checking though.

**Pull request merged**